### PR TITLE
Minimum Cost Path in Java

### LINKED ISSUE #4001: Minimum Cost Path
## 💥 Proposal

For an integer matrix of size (M x N), find out the minimum cost to reach from the cell (0, 0) to (M - 1, N - 1) in Java.

### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#reporting-new-issues)?

Yes

### PR DESCRIPTION
**Pull request by Ria700**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

### Description

 Find out the minimum cost to reach from the cell (0, 0) to (m - 1, n - 1) for a given m x n array.

### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.
- [x] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

Fixes #4001

### CODE REVIEW

**Comment by Ria700** [comment]
Added comments 
Changed Readme.md
Added a beginner-friendly description
Used proper code format
Removed extra lines
Added sample I/O
Added EOF

Please review

File: Java/dp/MinimumCostPath.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+                mat[row][col] = Integer.parseInt(strNums[col]);
+            }
+        }
+        return mat;
+    }
+//    Main function
+  public static void main(String[] args) throws NumberFormatException, IOException {
+        int[][] mat = take2DInput();
+        System.out.println("Solution Dynamic Programming:" + minCostPathDP(mat));
+        System.out.println("Solution Recursion:" + minCostPathRecursion(mat));
```
Comment (ankitaggarwal23) [inline Java/dp/MinimumCostPath.java]: Why you are using recursion and you mentioned through DP.

Comment (Ria700) [reply Java/dp/MinimumCostPath.java]: Added it for a better explanation.

**Comment by Ria700** [comment]
Removed Recursion Code and made the necessary changes. Please review!

File: Java/dp/MinimumCostPath.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
+// Description: Find out the minimum cost to reach from the cell (0, 0) to (m - 1, n - 1) for a given m x n array.
```
Comment (ankitaggarwal23) [inline Java/dp/MinimumCostPath.java]: - Add Multi-Line Comments 
- Add a beginner-friendly description at the top

File: Java/dp/MinimumCostPath.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+public class Solution {
+//    Dynamic Programming function
```
Comment (ankitaggarwal23) [inline Java/dp/MinimumCostPath.java]: Remove redundant comments

File: Java/dp/MinimumCostPath.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    public static int minCostPathDP(int[][] input) {
+    	int m = input.length;
+        int n = input[0].length;
```
Comment (ankitaggarwal23) [inline Java/dp/MinimumCostPath.java]: Add a tab space and align it

File: Java/dp/MinimumCostPath.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        int storage[][] = new int[m][n]; // storage array
+//       Filling up storage array
+//       Base cases
```
Comment (ankitaggarwal23) [inline Java/dp/MinimumCostPath.java]: Remove redundant comments

File: Java/dp/MinimumCostPath.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        storage[m-1][n-1] = input[m-1][n-1];
+        // Last row
+        for(int j = n-2; j >= 0; j--) {
+        	storage[m-1][j] = storage[m-1][j+1] + input[m-1][j];
+        }
+        // Last column
+        for(int i = m-2; i >=0; i--) {
+        	storage[i][n-1] = storage[i+1][n-1] + input[i][n-1];
+        }
+//        Complete array
```
Comment (ankitaggarwal23) [inline Java/dp/MinimumCostPath.java]: Remove redundant comments

File: Java/dp/MinimumCostPath.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        for(int i = m-2; i>=0; i--) {
+         	for(int j = n-2; j >= 0; j--) {
+        		storage[i][j] = input[i][j] + Math.min(storage[i][j+1], Math.min(storage[i+1][j+1], storage[i+1][j]));
+        	}
+        }
+        return storage[0][0];
+    }
+//    Take input
```
Comment (ankitaggarwal23) [inline Java/dp/MinimumCostPath.java]: Remove redundant comments

File: Java/dp/MinimumCostPath.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+            String[] strNums; 
+            strNums = br.readLine().trim().split("\\s");
+            for (int col = 0; col < nCols; col++) {
+                mat[row][col] = Integer.parseInt(strNums[col]);
+            }
+        }
+        return mat;
+    }
+//    Main function
+  public static void main(String[] args) throws NumberFormatException, IOException {
```
Comment (ankitaggarwal23) [inline Java/dp/MinimumCostPath.java]:
```
  }

  //Drive Code
  public static void main(String[] args) throws NumberFormatException, IOException {
```
Rewrite it like that

File: Java/dp/MinimumCostPath.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+  }
+}
+/*
+Sample Input 1 :
+3 4
+3 4 1 2
+2 1 8 9
+4 7 8 1
+Sample Output 1 :
+13
```
Comment (ankitaggarwal23) [inline Java/dp/MinimumCostPath.java]: Use proper display message in your code how your reading inputs and displaying outputs

File: Java/dp/MinimumCostPath.java [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+Sample Input 2 :
+3 4
+10 6 9 0
+-23 8 9 90
+-200 0 89 200
+Sample Output 2 :
+76
```
Comment (ankitaggarwal23) [inline Java/dp/MinimumCostPath.java]: Remove extra test case

File: Java/dp/MinimumCostPath.java [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+
+Time Complexity:
+Dynamic Programming: O(n^2)
+  
+Space Complexity:
+Dynamic Programming: O(m*n)
```
Comment (ankitaggarwal23) [inline Java/dp/MinimumCostPath.java]: Rewrite it like that,
```
Time Complexity: O(n^2)  
Space Complexity: O(m*n)
```

**Review by ankitaggarwal23 — CHANGES_REQUESTED** [review]

**Comment by Ria700** [comment]
Made all requested changes. Please review!

File: Java/dp/MinimumCostPath.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
+/* Description: An integer matrix of size (M x N) has been given. Find out the minimum cost to reach from the cell (0, 0) to (M - 1, N - 1).
+The cost of a path is defined as the sum of each cell's values through which the route passes.
+*/
```
Comment (ankitaggarwal23) [inline Java/dp/MinimumCostPath.java]:
```
/* 
Description: An integer matrix of size (M x N) has been given. Find 
out the minimum cost to reach from the cell (0, 0) to (M - 1, N - 1).
The cost of a path is defined as the sum of each cell's values through
which the route passes.
*/
```

Rewrite it like that

File: Java/dp/MinimumCostPath.java [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+        System.out.println("Solution:" + minCostPathDP(mat));
+  }
+}
+/*
+Input format :
+The first line of the test case contains two integer values, 'M' and 'N', separated by a single space. They represent the 'rows' and 'columns' respectively, for the two-dimensional array/list.
+The second line onwards, the next 'M' lines or rows represent the ith row values.
+Each of the ith row constitutes 'N' column values separated by a single space.
+Output format :
+Print the minimum cost to reach the destination.
```
Comment (ankitaggarwal23) [inline Java/dp/MinimumCostPath.java]: Remove this part

File: Java/dp/MinimumCostPath.java [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+Sample Input :
+3 4
+3 4 1 2
+2 1 8 9
+4 7 8 1
+Sample Output :
+13
```
Comment (ankitaggarwal23) [inline Java/dp/MinimumCostPath.java]: Use proper display message in your code how you are reading and writing outputs

**Review by ankitaggarwal23 — CHANGES_REQUESTED** [review]

**Comment by Ria700** [comment]
Made all requested changes. Please review!

File: Java/dp/MinimumCostPath.java [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+        int[][] mat = take2DInput();
+        System.out.println("Minimum cost to reach the destination: " + minCostPathDP(mat));
+  }
+}
+/*
+Sample Input :
+3 4
+3 4 1 2
+2 1 8 9
+4 7 8 1
```
Comment (ankitaggarwal23) [inline Java/dp/MinimumCostPath.java]: remove this

**Review by ankitaggarwal23 — CHANGES_REQUESTED** [review]

File: Java/dp/MinimumCostPath.java (hunk 15 above)
Comment (Ria700) [reply Java/dp/MinimumCostPath.java]: Done. Please Review!

**Review by ankitaggarwal23 — APPROVED** [review]

**Pull request merged**